File: lib/data/database/daos/event_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'event_dao.g.dart';

@DriftAccessor(tables: [CustomerEvents])
class EventDao extends DatabaseAccessor<AppDatabase> with _$EventDaoMixin {
  EventDao(super.db);

  Stream<List<CustomerEventEntity>> watchByCustomer(String customerId) =>
      (select(customerEvents)
            ..where((t) => t.customerId.equals(customerId))
            ..orderBy([(t) => OrderingTerm(expression: t.eventTime, mode: OrderingMode.desc)]))
          .watch();

  Future<List<CustomerEventEntity>> getByCustomer(String customerId) =>
      (select(customerEvents)
            ..where((t) => t.customerId.equals(customerId))
            ..orderBy([(t) => OrderingTerm(expression: t.eventTime, mode: OrderingMode.desc)]))
          .get();

  /// 查询某客户当天是否有某类型事件
  Future<bool> hasEventToday(String customerId, String eventType, DateTime date) async {
    final start = DateTime(date.year, date.month, date.day);
    final end = start.add(const Duration(days: 1));
    final query = select(customerEvents)
      ..where((t) =>
          t.customerId.equals(customerId) &
          t.eventType.equals(eventType) &
          t.eventTime.isBiggerOrEqualValue(start) &
          t.eventTime.isSmallerThanValue(end));
    final result = await query.get();
    return result.isNotEmpty;
  }

  /// 统计当天某类型事件数量
  Future<int> countEventToday(String eventType, DateTime date) async {
    final start = DateTime(date.year, date.month, date.day);
    final end = start.add(const Duration(days: 1));
    final countExp = customerEvents.eventType.count();
    final query = selectOnly(customerEvents)
      ..addColumns([countExp])
      ..where(customerEvents.eventType.equals(eventType) &
          customerEvents.eventTime.isBiggerOrEqualValue(start) &
          customerEvents.eventTime.isSmallerThanValue(end));
    final result = await query.map((row) => row.read(countExp) ?? 0).getSingle();
    return result;
  }

  /// 监听当天某类型事件数量 (Stream)
  Stream<int> watchCountEventToday(String eventType, DateTime date) {
    final start = DateTime(date.year, date.month, date.day);
    final end = start.add(const Duration(days: 1));
    final countExp = customerEvents.eventType.count();
    final query = selectOnly(customerEvents)
      ..addColumns([countExp])
      ..where(customerEvents.eventType.equals(eventType) &
          customerEvents.eventTime.isBiggerOrEqualValue(start) &
          customerEvents.eventTime.isSmallerThanValue(end));
    return query.map((row) => row.read(countExp) ?? 0).watchSingle();
  }

  /// 统计某类型事件在时间范围内的数量
  Future<int> countEventRange(String eventType, DateTime start, DateTime end) async {
    final countExp = customerEvents.eventType.count();
    final query = selectOnly(customerEvents)
      ..addColumns([countExp])
      ..where(customerEvents.eventType.equals(eventType) &
          customerEvents.eventTime.isBiggerOrEqualValue(start) &
          customerEvents.eventTime.isSmallerThanValue(end));
    final result = await query.map((row) => row.read(countExp) ?? 0).getSingle();
    return result;
  }

  /// 统计某类型事件总数
  Future<int> countEventTotal(String eventType) async {
    final countExp = customerEvents.eventType.count();
    final query = selectOnly(customerEvents)
      ..addColumns([countExp])
      ..where(customerEvents.eventType.equals(eventType));
    final result = await query.map((row) => row.read(countExp) ?? 0).getSingle();
    return result;
  }

  Future<String> insertEvent(CustomerEventsCompanion entry) async {
    final id = entry.id.present ? entry.id.value : _genId();
    await into(customerEvents).insert(entry.copyWith(id: Value(id)));
    return id;
  }

  String _genId() {
    final now = DateTime.now();
    return 'evt_${now.microsecondsSinceEpoch}';
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/EventDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.salesquest.sales_quest.data.entity.CustomerEventEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface EventDao {

    @Query("SELECT * FROM customer_events WHERE customerId = :customerId ORDER BY eventTime DESC")
    fun watchByCustomer(customerId: String): Flow<List<CustomerEventEntity>>

    @Query("SELECT * FROM customer_events WHERE customerId = :customerId ORDER BY eventTime DESC")
    suspend fun getByCustomer(customerId: String): List<CustomerEventEntity>

    @Query("SELECT * FROM customer_events")
    suspend fun getAll(): List<CustomerEventEntity>

    @Query(
        "SELECT * FROM customer_events WHERE customerId = :customerId AND eventType = :eventType " +
            "AND eventTime >= :start AND eventTime < :end"
    )
    suspend fun getEventsToday(customerId: String, eventType: String, start: Long, end: Long): List<CustomerEventEntity>

    @Query("SELECT COUNT(*) FROM customer_events WHERE eventType = :eventType AND eventTime >= :start AND eventTime < :end")
    suspend fun countEventToday(eventType: String, start: Long, end: Long): Int

    @Query("SELECT COUNT(*) FROM customer_events WHERE eventType = :eventType AND eventTime >= :start AND eventTime < :end")
    fun watchCountEventToday(eventType: String, start: Long, end: Long): Flow<Int>

    @Query("SELECT COUNT(*) FROM customer_events WHERE eventType = :eventType AND eventTime >= :start AND eventTime < :end")
    suspend fun countEventRange(eventType: String, start: Long, end: Long): Int

    @Query("SELECT COUNT(*) FROM customer_events WHERE eventType = :eventType")
    suspend fun countEventTotal(eventType: String): Int

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertEvent(event: CustomerEventEntity)

    @Query("DELETE FROM customer_events")
    suspend fun clearAll()
}
